File: managed/node-agent/resources/ynp/configs/setup_logger.py
import logging
import logging.config
import os
import pwd
import time


def setup_logger(config):
    key = next(iter(config), None)
    log_file = "app.log"
    log_dir = "./logs"
    log_level = "DEBUG"
    if key is not None:
        context = config[key]
        log_file = context.get('logfile', log_file)
        log_dir = context.get("logdir", log_dir)
        log_level = context.get('loglevel', log_level).upper()

    # Create log directory if it doesn't exist
    # Ensure the log directory has the correct permissions (755)
    # We need execute set on directory for traversal, don't need it on all files.
    os.makedirs(log_dir, mode=0o755, exist_ok=True)

    log_path = os.path.join(log_dir, log_file)
    logging_config = {
        'version': 1,
        'disable_existing_loggers': False,
        'formatters': {
            'standard': {
                'format': '%(asctime)s - %(name)s - %(levelname)s - %(message)s',
                'datefmt': '%Y-%m-%dT%H:%M:%SZ',  # ISO-like UTC format
                '()': 'logging.Formatter',        # Explicit formatter class
            },
        },
        'handlers': {
            'console': {
                'class': 'logging.StreamHandler',
                'formatter': 'standard',
                'level': log_level,
            },
            'file': {
                'class': 'logging.FileHandler',
                'formatter': 'standard',
                'level': log_level,
                'filename': log_path,
                'mode': 'a',
            },
        },
        'loggers': {
            '': {  # root logger
                'handlers': ['console', 'file'],
                'level': log_level,
                'propagate': True,
            }
        }
    }
    logging.config.dictConfig(logging_config)
    # Patch all formatters to use UTC time
    for handler in logging.getLogger().handlers:
        if isinstance(handler.formatter, logging.Formatter):
            handler.formatter.converter = time.gmtime

    # Set file permissions (644)
    os.chmod(log_path, 0o644)
    current_user_id = os.getuid()
    # Set ownership to original user if script was run with sudo
    if 'SUDO_USER' in os.environ:
        original_user = os.environ['SUDO_USER']
    else:
        original_user = os.getlogin()
    user_info = pwd.getpwnam(original_user)
    uid = user_info.pw_uid
    gid = user_info.pw_gid
    if current_user_id == 0 and current_user_id != uid:
        # Change ownership only if running as root and yb_user is different from current user.
        os.chown(log_dir, uid, gid)
        os.chown(log_path, uid, gid)
    logger = logging.getLogger()
    logger.info(f"Logging setup complete in UTC timezone. Level: {log_level}, Log File: {log_path}")

File: managed/node-agent/ynp/config/setup_logger.go
// Copyright (c) YugabyteDB, Inc.

package config

import (
	"context"
	"log"
	"node-agent/util"
	"os"
	"os/user"
	"path/filepath"

	apex "github.com/apex/log"
)

const (
	DefaultLogFile    = "app.log"
	DefaultLogDir     = "./logs"
	DefaultMaxSizeMB  = 10
	DefaultMaxBackups = 5
	DefaultMaxAgeDays = 30
)

func SetupLogger(ctx context.Context, config map[string]map[string]any) {
	// Default values
	logFile := DefaultLogFile
	logDir := DefaultLogDir
	logLevel := apex.DebugLevel
	logging, ok := config["logging"]
	if ok {
		if lf, ok := logging["file"].(string); ok && lf != "" {
			logFile = lf
		}
		if ld, ok := logging["directory"].(string); ok && ld != "" {
			logDir = ld
		}
		if ll, ok := logging["level"].(string); ok && ll != "" {
			level, err := apex.ParseLevel(ll)
			if err == nil {
				logLevel = level
			} else {
				log.Printf("Invalid log level '%s', defaulting to 'info'", ll)
			}
		}
	}
	// Create log directory if it doesn't exist
	err := os.MkdirAll(logDir, 0755)
	if err != nil {
		log.Fatalf("Failed to create log directory: %v", err)
	}
	logPath := filepath.Join(logDir, logFile)
	util.InitCustomAppLogger(
		logPath,
		DefaultMaxSizeMB,
		DefaultMaxBackups,
		DefaultMaxAgeDays,
		logLevel,
		true,  /* enableConsole */
		false, /* loadConfigFile */
	)
	// Set file permissions
	_ = os.Chmod(logPath, 0644)
	// Set ownership to original user if run with sudo
	origUser := os.Getenv("SUDO_USER")
	if origUser == "" {
		if u, err := user.Current(); err == nil {
			origUser = u.Username
		}
	}
	if origUser != "" {
		if details, err := util.UserInfo(origUser); err == nil {
			if details.UserID == 0 {
				_ = os.Chown(logDir, int(details.UserID), int(details.GroupID))
				_ = os.Chown(logPath, int(details.UserID), int(details.GroupID))
			}
		}
	}
	util.FileLogger().Infof(ctx, "Logging setup complete in UTC timezone. Level: %s, path: %s",
		logLevel.String(), logPath)
}
